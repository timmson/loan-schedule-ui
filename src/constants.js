export const CHANGE_FORM = "CHANGE_FORM"
export const CUT_SCHEDULE = "CUT_SCHEDULE"
export const INIT = "INIT"
export const SET_DEFAULT = "SET_DEFAULT"
export const UPDATE_SCHEDULE = "UPDATE_SCHEDULE"
