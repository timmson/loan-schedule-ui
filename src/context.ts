import {createContext, Dispatch} from "react"
import {ActionType} from "./types"

const defaultValue: Dispatch<ActionType> = () => {}
export default createContext(defaultValue)