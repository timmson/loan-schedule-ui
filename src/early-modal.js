import React from "react"
import PropTypes from "prop-types"
import {Button, Modal} from "react-bootstrap"

export default function EarlyModal(props) {
	return (
		<Modal show={props.show}>
			<Modal.Header closeButton onClick={() => props.close.action()}>
				<Modal.Title>{props.name}</Modal.Title>
			</Modal.Header>

			<Modal.Body>
				{props.children}
			</Modal.Body>

			<Modal.Footer>
				<Button variant="secondary" onClick={() => props.close.action()}>{props.close.name}</Button>
				<Button style={{background: "#fd680e", border: "#fd680e"}} onClick={() => props.ok.action()}>{props.ok.name}</Button>
			</Modal.Footer>
		</Modal>
	)
}

EarlyModal.propTypes = {
	show: PropTypes.bool.isRequired,
	name: PropTypes.string.isRequired,
	ok: PropTypes.object.isRequired,
	close: PropTypes.object.isRequired,
	children: PropTypes.any
}

