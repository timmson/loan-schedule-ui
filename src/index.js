import "bootstrap"
import "./index.scss"

import React from "react"
import ReactDOM from "react-dom"
import LoanSchedule from "loan-schedule.js"
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import {faTelegram} from "@fortawesome/free-brands-svg-icons"

import App from "./app"
import Storage from "./storage"
import ReducerFactory from "./reducer-factory"

const storage = Storage(window)
const reducer = ReducerFactory(storage, new LoanSchedule({prodCalendar: "ru"})).reducer

const currentYear = new Date().getFullYear()
const shareUrl = () => {
	window.open(`https://t.me/share/url?url=${encodeURIComponent(window.location.href)}`, "blank")
}

ReactDOM.render(
	<div className="container">
		<App reducer={reducer}/>
		<div className="row mt-5">
			<div className="col text-end">
				<a href="#" target="_blank" onClick={() => shareUrl()}>
					[Share via <FontAwesomeIcon icon={faTelegram}/>]
				</a>
				<p className="copyright">&copy; {currentYear} timmson</p>
			</div>
			<div className="col-sm-1">&nbsp;</div>
		</div>
	</div>
	, document.getElementById("app"))


/*const telegramShareUrl = "https://t.me/share/url"*/
/*new Vue({
	el: "#app",
	data: {
		currentYear: new Date().getFullYear().toString(),
		shareUrl: "",
		request: storage.load(),
		earlyRepayment: {
			date: "",
			amount: ""
		},
		schedule: {}
	},
	methods: {

		updateSchedule: function () {

			this.request.amount = this.fromMoney(this.request.amount);
			this.request.paymentAmount = this.fromMoney(this.request.paymentAmount);
			this.request.paymentOnDay = parseInt(this.request.paymentOnDay) > 31 ? 31 : this.request.paymentOnDay;

			if (this.earlyRepayment.date) {
				delete this.request.earlyRepayment[this.earlyRepayment.date];
				if (this.earlyRepayment.amount) {
					this.request.earlyRepayment[this.earlyRepayment.date] = {
						"erAmount": this.fromMoney(this.earlyRepayment.amount),
						"erType": LoanSchedule.ER_TYPE_MATURITY
					};
				}
			}

			this.schedule = loanSchedule.calculateSchedule(this.request);
			this.schedule.lastPaymentDate = this.schedule.payments[this.schedule.payments.length - 1].paymentDate;
			this.schedule.termInYear = Math.ceil(this.schedule.term / 12);

			storage.save(this.request);
			this.shareUrl = `${telegramShareUrl}?url=${encodeURIComponent(window.location.href)}`;

			this.request.amount = this.toMoney(this.request.amount);
			this.request.paymentAmount = this.toMoney(this.request.paymentAmount);
			this.earlyRepayment.amount = this.toMoney(this.earlyRepayment.amount);
		},

		copyPayment: function (event, paymentId) {
			this.request.amount = this.schedule.payments[paymentId - 1].finalBalance;
			this.request.term = this.request.term - paymentId + 1;
			this.request.issueDate = this.schedule.payments[paymentId - 1].paymentDate;
			this.request.earlyRepayment = {};
			this.updateSchedule(null);
		},

		toMoney: function (number) {
			return number ? Accounting.formatMoney(number, {symbol: "", format: "%s%v", thousand: " "}) : number;
		},

		fromMoney: function (numberWithSpaces) {
			return numberWithSpaces ? numberWithSpaces.toString().split(" ").join("").replaceAll(",", ".") : numberWithSpaces;
		},

		reset: function () {
			storage.reset();
			window.location = ".";
		}
	},

	mounted() {
		this.updateSchedule(null);
	}
});*/

