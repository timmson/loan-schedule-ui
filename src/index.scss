@import "~bootstrap/scss/bootstrap.scss";

.orange {
  color: #fd680e;
}

.table-bordered {
  font-size: 0.8em;
}

.desktop {
  display: table-row;
}

.desktop-cell {
  display: table-cell;
}


.mobile {
  display: none;
}

@media (-webkit-min-device-pixel-ratio: 2) {
  html, body {
    font-size: 14pt;
  }

  .desktop {
    display: none;
  }

  .desktop-cell {
    display: none;
  }

  .mobile {
    display: table-row;
  }
}