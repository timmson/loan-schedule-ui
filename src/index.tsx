import "bootstrap"
import "./index.scss"

import React from "react"
import {createRoot} from "react-dom/client"
import LoanSchedule from "loan-schedule.js"
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import {faTelegram} from "@fortawesome/free-brands-svg-icons"

import App from "./app"
import Storage from "./storage"
import Reducer from "./reducer"

const storage = Storage(window)
const reducer = Reducer(storage, new LoanSchedule({prodCalendar: "ru"}))

const currentYear = new Date().getFullYear()
const shareUrl = () => {
	window.open(`https://t.me/share/url?url=${encodeURIComponent(window.location.href)}`, "blank")
}

const container = document.getElementById("app")

createRoot(container!).render(
	<div className="container" style={{marginTop: 0, paddingTop: 0}}>
		<App reducer={reducer}/>
		<div className="row mt-3">
			<div className="col text-end">
				<a href="#" target="_blank" onClick={() => shareUrl()}>
                    [Share via <FontAwesomeIcon icon={faTelegram}/>]
				</a>
				<p className="copyright">&copy; {currentYear} timmson</p>
			</div>
			<div className="col-sm-1">&nbsp;</div>
		</div>
	</div>
)
