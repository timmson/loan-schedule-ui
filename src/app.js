import React, {useEffect, useReducer} from "react"
import PropTypes from "prop-types"
import {faTelegram} from "@fortawesome/free-brands-svg-icons"
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"

import Context from "./context"
import Form from "./form"
import {toM} from "./money"
import Reducer from "./reducer"
import Schedule from "./schedule"
import {UPDATE_SCHEDULE} from "./constants"
import {faArrowAltCircleDown, faArrowAltCircleUp} from "@fortawesome/free-regular-svg-icons"

const shareUrl = () => {
	window.open(`https://t.me/share/url?url=${encodeURIComponent(window.location.href)}`, "blank")
}


export default function App(props) {
	const currentYear = new Date().getFullYear()

	const request = props.storage.load()
	const [state, dispatch] = useReducer(Reducer, {}, () => Reducer({request: request}, {type: UPDATE_SCHEDULE}))

	useEffect(() => props.storage.save(state.request), [state.request])

	return (
		<Context.Provider value={dispatch}>
			<div className="container">
				<Form request={state.request}/>

				<div className="row text-start mt-1">
					<div className="col orange">
						{toM(state.schedule.overAllInterest)} ({toM(state.schedule.fullAmount)}),
						{state.schedule.lastPaymentDate} (~{state.schedule.termInYear}Y)
					</div>
				</div>
				<div className="row text-end mt-1">
					<div className="col">
						<a href="#">[Reset]</a>
						<a href="#footer" id="header">[Bottom <FontAwesomeIcon icon={faArrowAltCircleDown}/>]</a>
					</div>
				</div>

				<Schedule schedule={state.schedule}/>

				<div className="row">
					<div className="col-12 text-end ps-2">
						<a href="#header" id="footer">[Top <FontAwesomeIcon icon={faArrowAltCircleUp}/>]</a>
					</div>
				</div>
				<div className="row mt-5">
					<div className="col text-end">
						<a href="#" target="_blank" onClick={() => shareUrl()}>
							[Share via <FontAwesomeIcon icon={faTelegram}/>]
						</a>
						<p className="copyright">&copy; {currentYear} timmson</p>
					</div>
					<div className="col-sm-1">&nbsp;</div>
				</div>
			</div>
		</Context.Provider>
	)
}

App.propTypes = {
	storage: PropTypes.object.isRequired
}