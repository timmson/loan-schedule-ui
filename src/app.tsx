import React, {useReducer} from "react"
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import {faArrowAltCircleDown, faArrowAltCircleUp} from "@fortawesome/free-regular-svg-icons"

import {INIT, SET_DEFAULT} from "./constants"
import Context from "./context"
import Form from "./form"
import Schedule from "./schedule"
import {ActionType, StateType} from "./types"

type AppProps = {
	reducer: (state: StateType, action: ActionType) => StateType
}

export default function App(props: AppProps) {
	const [state, dispatch] = useReducer(props.reducer, {}, () => props.reducer({}, {type: INIT}))
	const reset = () => dispatch({type: SET_DEFAULT})

	return (
		<Context.Provider value={dispatch}>
			<Form request={state.request}/>
			<div className="row text-start mt-1 orange border-bottom border-2 border-dark pb-2">
				<div className="col-lg-4 orange">
					<b>Переплата, ₽</b>: {state.schedule.overAllInterest}
				</div>
				<div className="col-lg-4 orange">
					<b>Полная стоимость, ₽</b>: {state.schedule.fullAmount}
				</div>
				<div className="col-lg-4 orange">
					<b>Последний платеж</b>: {state.schedule.lastPaymentDate} (~{state.schedule.termInYear} год(а)/лет)
				</div>
			</div>
			<div className="row text-end mt-1">
				<div className="col">
					<a href="#" onClick={() => reset()}>[Reset]</a>
					<a href="#footer" id="header">[Bottom <FontAwesomeIcon icon={faArrowAltCircleDown}/>]</a>
				</div>
			</div>

			<Schedule schedule={state.schedule}/>

			<div className="row">
				<div className="col-12 text-end ps-2">
					<a href="#header" id="footer">[Top <FontAwesomeIcon icon={faArrowAltCircleUp}/>]</a>
				</div>
			</div>
		</Context.Provider>
	)
}
