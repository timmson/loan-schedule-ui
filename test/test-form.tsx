import React from "react"
import renderer from "react-test-renderer"
import Context from "../src/context"
import {CHANGE_FORM, UPDATE_SCHEDULE} from "../src/constants"

import Form from "../src/form"

describe("Form should", () => {

	let component = null

	const createComponent = (request, dispatch) => renderer.create(
		<Context.Provider value={dispatch}>
			<Form request={request}/>
		</Context.Provider>
	)

	test("trigger blur", () => {
		const dispatch = (action) => expect(action).toEqual({type: UPDATE_SCHEDULE})
		component = createComponent({}, dispatch)

		component.root.findByType("form").props.onBlur()

		expect.assertions(1)
	})

	const tc = [
		{name: "amount", value: "value"},
		{name: "paymentAmount", value: "value"},
		{name: "term", value: "value"},
		{name: "issueDate", value: "value"},
		{name: "rate", value: "value"},
		{name: "paymentOnDay", value: "value"}
	]
	tc.forEach((t) =>
		test(`trigger change of ${t.name} with value=${t.value}`, () => {
			const expectedAction = {type: CHANGE_FORM, name: t.name, value: t.value}
			const dispatch = (action) => expect(action).toEqual(expectedAction)
			component = createComponent({}, dispatch)

			const event = {target: {name: t.name, value: t.value}}
			component.root.findByProps({"name": t.name}).props.onChange(event)

			expect.assertions(1)
		})
	)


	test("equal to snapshot", () => {
		component = createComponent({}, () => null)
		expect(component.toJSON()).toMatchSnapshot()
	})

	afterAll(() => {
		component.unmount()
	})

})