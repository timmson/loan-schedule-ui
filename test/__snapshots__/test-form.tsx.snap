// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should equal to snapshot 1`] = `
<form
  className="form-horizontal"
  onBlur={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2
        className="orange text-center"
      >
        График погашения кредита
      </h2>
    </div>
  </div>
  <div
    className="row mt-2 text-left"
  >
    <div
      className="col-lg-3 pt-2"
    >
      <label
        htmlFor="amount"
      >
        Сумма кредита, ₽
      </label>
    </div>
    <div
      className="col-lg-3"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
      />
    </div>
    <div
      className="col-lg-3 pt-2"
    >
      <label
        htmlFor="paymentAmount"
      >
        Сумма платежа, ₽
      </label>
    </div>
    <div
      className="col-lg-3"
    >
      <input
        className="form-control"
        name="paymentAmount"
        onChange={[Function]}
        placeholder="Сумма платежа, ₽"
      />
    </div>
  </div>
  <div
    className="row mt-2 text-left"
  >
    <div
      className="col-lg-3 pt-2"
    >
      <label
        htmlFor="term"
      >
        Срок, месяцев
      </label>
    </div>
    <div
      className="col-lg-3"
    >
      <input
        className="form-control"
        name="term"
        onChange={[Function]}
        type="number"
      />
    </div>
    <div
      className="col-lg-3 pt-2"
    >
      <label
        htmlFor="issueDate"
      >
        Дата выдача
      </label>
    </div>
    <div
      className="col-lg-3"
    >
      <input
        className="form-control"
        name="issueDate"
        onChange={[Function]}
        placeholder="dd.mm.yyyy"
      />
    </div>
  </div>
  <div
    className="row mt-2 text-left border-bottom border-2 border-dark pb-2"
  >
    <div
      className="col-lg-3 pt-2"
    >
      <label
        htmlFor="rate"
      >
        Ставка, %
      </label>
    </div>
    <div
      className="col-lg-3"
    >
      <input
        className="form-control"
        name="rate"
        onChange={[Function]}
        step="0.01"
        type="number"
      />
    </div>
    <div
      className="col-lg-3 pt-2"
    >
      <label
        htmlFor="paymentOnDay"
      >
        День платежа
      </label>
    </div>
    <div
      className="col-lg-3"
    >
      <input
        className="form-control"
        max="28"
        min="0"
        name="paymentOnDay"
        onChange={[Function]}
      />
    </div>
  </div>
</form>
`;
